### PR TITLE
python overload vstack

### PR DESCRIPTION
**Pull request by ritchie46**
Overload vstack #1984

### CODE REVIEW

**Comment by MarcoGorelli** [comment]
Just FYI, you might want to add an extra one for when you don't know the value of `in_place`, as currently, if it's just a bool handed down by another function, you get:
```python
import polars
from typing import Literal, Union

df: polars.DataFrame
in_place: bool

reveal_type(df.vstack(df, in_place=True))
reveal_type(df.vstack(df, in_place=False))
reveal_type(df.vstack(df, in_place=in_place))
```

```console
$ mypy t.py
$ mypy t.py 
t.py:7: note: Revealed type is "None"
t.py:8: note: Revealed type is "polars.internals.frame.DataFrame"
t.py:9: error: No overload variant of "vstack" of "DataFrame" matches argument types "DataFrame", "bool"
t.py:9: note: Possible overload variants:
t.py:9: note:     def vstack(self, df: DataFrame, in_place: Literal[True]) -> None
t.py:9: note:     def vstack(self, df: DataFrame, in_place: Literal[False] = ...) -> DataFrame
t.py:9: note: Revealed type is "Any"
Found 1 error in 1 file (checked 1 source file)
```

File: py-polars/polars/internals/frame.py [hunk 1]
```diff
@@ -2493,6 +2493,14 @@ def hstack(
         else:
             return wrap_df(self._df.hstack([s.inner() for s in columns]))
 
+    @tp.overload
+    def vstack(self, df: "DataFrame", in_place: Literal[True]) -> None:
+        ...
+
+    @tp.overload
+    def vstack(self, df: "DataFrame", in_place: Literal[False] = ...) -> "DataFrame":
+        ...
```
Comment (MarcoGorelli) [inline py-polars/polars/internals/frame.py:2502]:
```suggestion
    @tp.overload
    def vstack(self, df: "DataFrame", in_place: Literal[False] = ...) -> "DataFrame":
        ...


    @tp.overload
    def vstack(self, df: "DataFrame", in_place: bool) -> Optional["DataFrame"]:
        ...
```

Comment (MarcoGorelli) [reply py-polars/polars/internals/frame.py:2502]: If I add this in, then the snipped posted above gives:
```console
$ mypy t.py 
t.py:7: note: Revealed type is "None"
t.py:8: note: Revealed type is "polars.internals.frame.DataFrame"
t.py:9: note: Revealed type is "Union[polars.internals.frame.DataFrame, None]"
```

Comment (zundertj) [reply py-polars/polars/internals/frame.py:2502]: I think this is a reasonable compromise for bools / in-place, it does add some clutter to the code base having to write three overloads per method though.

Note that similar issues with strings have been raised before: https://github.com/pola-rs/polars/pull/1937#discussion_r759971391, where unfortunately the fall back means any string is accepted.

**Comment by ritchie46** [comment]
> It's frustrating, but if you have overloads, then mypy no longer looks at the original signature itself, it just tries to match one of the overloads. 

Ah.. clear. Thanks!

File: py-polars/polars/internals/frame.py (hunk 1 above)
Comment (ritchie46) [reply py-polars/polars/internals/frame.py:2502]: > I think this is a reasonable compromise for bools / in-place, it does add some clutter to the code base having to write three overloads per method though.

Yes, hopefully we don't have too many return type changing arguments.

**Pull request merged**